File: code_examples/WifiP2pManagerSingleton.java
/**
 * WiFiP2P Manager Singleton - Manages WiFiP2P device discovery and connections
 * File: heycyan_rev/_log_review/app/src/main/java/sources/com/glasssutdio/wear/wifi/p2p/WifiP2pManagerSingleton.java
 */

import android.app.Application;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.net.wifi.p2p.WifiP2pConfig;
import android.net.wifi.p2p.WifiP2pDevice;
import android.net.wifi.p2p.WifiP2pInfo;
import android.net.wifi.p2p.WifiP2pManager;
import android.os.Handler;
import android.os.Looper;
import com.elvishew.xlog.XLog;
import com.oudmon.ble.base.communication.ILargeDataResponse;
import com.oudmon.ble.base.communication.LargeDataHandler;
import com.oudmon.ble.base.communication.bigData.resp.BaseResponse;
import com.oudmon.ble.base.communication.bigData.resp.GlassModelControlResponse;

import java.util.Collection;

public final class WifiP2pManagerSingleton {
    
    public static final Companion INSTANCE = new Companion();
    private static volatile WifiP2pManagerSingleton instance;
    
    private WifiP2pCallback callback;
    private int connectRetry;
    private final ConnectTimeOut connectTimeOut;
    private boolean connected;
    private boolean connecting;
    private final Context context;
    private int discoveryRetry;
    private final DiscoveryTimeOut discoveryTimeOut;
    private final Handler handler;
    private final IntentFilter intentFilter;
    private final WifiP2pBroadcastReceiver receiver;
    private WifiP2pManager.Channel wifiP2pChannel;
    private WifiP2pDevice wifiP2pDevice;
    private WifiP2pManager wifiP2pManager;
    
    /**
     * WiFiP2P Callback Interface
     */
    public interface WifiP2pCallback {
        void cancelConnect();
        void cancelConnectFail(int reason);
        void connecting();
        void onConnectRequestFailed(int reason);
        void onConnectRequestSent();
        void onConnected(WifiP2pInfo info);
        void onDisconnected();
        void onPeerDiscoveryFailed(int reason);
        void onPeerDiscoveryStarted();
        void onPeersChanged(Collection<? extends WifiP2pDevice> peers);
        void onThisDeviceChanged(WifiP2pDevice device);
        void onWifiP2pDisabled();
        void onWifiP2pEnabled();
        void retryAlsoFailed();
    }
    
    /**
     * Constructor
     */
    private WifiP2pManagerSingleton(Context context) {
        this.context = context;
        Object systemService = context.getSystemService("wifip2p");
        this.wifiP2pManager = (WifiP2pManager) systemService;
        
        this.wifiP2pChannel = wifiP2pManager.initialize(context, context.getMainLooper(), 
            new WifiP2pManager.ChannelListener() {
                @Override
                public void onChannelDisconnected() {
                    XLog.e("wifiP2pChannel disconnect");
                }
            });
            
        IntentFilter intentFilter = new IntentFilter();
        intentFilter.addAction(WifiP2pManager.WIFI_P2P_STATE_CHANGED_ACTION);
        intentFilter.addAction(WifiP2pManager.WIFI_P2P_PEERS_CHANGED_ACTION);
        intentFilter.addAction(WifiP2pManager.WIFI_P2P_CONNECTION_STATE_CHANGE_ACTION);
        intentFilter.addAction(WifiP2pManager.WIFI_P2P_THIS_DEVICE_CHANGED_ACTION);
        this.intentFilter = intentFilter;
        
        this.handler = new Handler(context.getMainLooper());
        this.discoveryTimeOut = new DiscoveryTimeOut(this);
        this.connectTimeOut = new ConnectTimeOut(this);
        this.receiver = new WifiP2pBroadcastReceiver(this);
    }
    
    /**
     * Public methods
     */
    public void registerReceiver() {
        this.context.registerReceiver(this.receiver, this.intentFilter);
    }
    
    public void unregisterReceiver() {
        this.context.unregisterReceiver(this.receiver);
    }
    
    public void resetPeerDiscovery() {
        this.handler.removeCallbacks(this.discoveryTimeOut);
    }
    
    public void resetFailCount() {
        this.connectRetry = 0;
        this.discoveryRetry = 0;
        this.connecting = false;
        setConnect(false);
        this.handler.removeCallbacks(this.discoveryTimeOut);
        this.handler.removeCallbacks(this.connectTimeOut);
    }
    
    public void startPeerDiscovery() {
        this.handler.postDelayed(this.discoveryTimeOut, 16000L);
        this.wifiP2pManager.discoverPeers(this.wifiP2pChannel, new WifiP2pManager.ActionListener() {
            @Override
            public void onSuccess() {
                if (callback != null) {
                    callback.onPeerDiscoveryStarted();
                }
            }
            
            @Override
            public void onFailure(int reason) {
                handler.removeCallbacks(discoveryTimeOut);
                handler.postDelayed(discoveryTimeOut, 2000L);
                if (callback != null) {
                    callback.onPeerDiscoveryFailed(reason);
                }
            }
        });
    }
    
    public void setConnect(boolean connected) {
        this.connected = connected;
    }
    
    public void connectToDevice(WifiP2pDevice device) {
        try {
            resetPeerDiscovery();
            if (this.connected) {
                XLog.i("P2P已经连接上了，直接返回");
                return;
            }
            if (this.connecting) {
                if (callback != null) {
                    callback.connecting();
                }
                XLog.i("P2P正在连接,不调用连接返回");
                return;
            }
            this.handler.postDelayed(this.connectTimeOut, 15000L);
            this.wifiP2pDevice = device;
            WifiP2pConfig wifiP2pConfig = new WifiP2pConfig();
            wifiP2pConfig.deviceAddress = device.deviceAddress;
            wifiP2pConfig.wps.setup = 0;
            this.connecting = true;
            XLog.i("已经在连接设备:" + device.deviceName);
            this.wifiP2pManager.connect(this.wifiP2pChannel, wifiP2pConfig, 
                new WifiP2pManager.ActionListener() {
                    @Override
                    public void onSuccess() {
                        if (callback != null) {
                            callback.onConnectRequestSent();
                        }
                    }
                    
                    @Override
                    public void onFailure(int reason) {
                        connecting = false;
                        if (callback != null) {
                            callback.onConnectRequestFailed(reason);
                        }
                    }
                });
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
    
    public void cancelP2pConnection() {
        try {
            initP2P();
            this.wifiP2pManager.cancelConnect(this.wifiP2pChannel, 
                new WifiP2pManager.ActionListener() {
                    @Override
                    public void onSuccess() {
                        if (callback != null) {
                            callback.cancelConnect();
                        }
                    }
                    
                    @Override
                    public void onFailure(int reason) {
                        if (callback != null) {
                            callback.cancelConnectFail(reason);
                        }
                    }
                });
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
    
    public void addCallback(WifiP2pCallback callback) {
        this.callback = callback;
    }
    
    public void removeCallback() {
        instance = null;
    }
    
    /**
     * Private helper methods
     */
    private void resetDeviceP2p() {
        LargeDataHandler.getInstance().glassesControl(new byte[]{2, 1, 15}, 
            new ILargeDataResponse() {
                @Override
                public void parseData(int cmdType, BaseResponse response) {
                    // Handle response
                }
            });
    }
    
    private void initP2P() {
        this.wifiP2pChannel.close();
        Object systemService = this.context.getSystemService("wifip2p");
        this.wifiP2pManager = (WifiP2pManager) systemService;
        this.wifiP2pChannel = wifiP2pManager.initialize(this.context, Looper.getMainLooper(), 
            new WifiP2pManager.ChannelListener() {
                @Override
                public void onChannelDisconnected() {
                    XLog.i("wifiP2pChannel initP2P");
                }
            });
    }
    
    /**
     * Inner classes
     */
    private static final class DiscoveryTimeOut implements Runnable {
        private final WifiP2pManagerSingleton outer;
        
        public DiscoveryTimeOut(WifiP2pManagerSingleton outer) {
            this.outer = outer;
        }
        
        @Override
        public void run() {
            XLog.i("内部扫描重试连接:" + this.outer.discoveryRetry);
            if (this.outer.discoveryRetry < 1) {
                XLog.i("内部扫描重试连接一次");
                this.outer.resetDeviceP2p();
                this.outer.initP2P();
                this.outer.startPeerDiscovery();
                this.outer.discoveryRetry++;
            }
        }
    }
    
    private static final class ConnectTimeOut implements Runnable {
        private final WifiP2pManagerSingleton outer;
        
        public ConnectTimeOut(WifiP2pManagerSingleton outer) {
            this.outer = outer;
        }
        
        @Override
        public void run() {
            this.outer.connecting = false;
            if (this.outer.connectRetry < 1) {
                WifiP2pDevice wifiP2pDevice = this.outer.wifiP2pDevice;
                if (wifiP2pDevice != null) {
                    XLog.i("内部连接重试连接一次");
                    this.outer.connectToDevice(wifiP2pDevice);
                }
                this.outer.connectRetry++;
            } else {
                XLog.i("不重连，等外部超时");
                if (this.outer.callback != null) {
                    this.outer.callback.retryAlsoFailed();
                }
            }
        }
    }
    
    private static final class WifiP2pBroadcastReceiver extends BroadcastReceiver {
        private final WifiP2pManagerSingleton outer;
        
        public WifiP2pBroadcastReceiver(WifiP2pManagerSingleton outer) {
            this.outer = outer;
        }
        
        @Override
        public void onReceive(Context context, Intent intent) {
            String action = intent.getAction();
            
            if (action == null) return;
            
            switch (action) {
                case WifiP2pManager.WIFI_P2P_STATE_CHANGED_ACTION:
                    int state = intent.getIntExtra(WifiP2pManager.EXTRA_WIFI_STATE, -1);
                    if (state == WifiP2pManager.WIFI_P2P_STATE_ENABLED) {
                        XLog.d("WiFi P2P enabled");
                        if (outer.callback != null) {
                            outer.callback.onWifiP2pEnabled();
                        }
                    } else {
                        XLog.d("WiFi P2P disabled");
                        if (outer.callback != null) {
                            outer.callback.onWifiP2pDisabled();
                        }
                    }
                    break;
                    
                case WifiP2pManager.WIFI_P2P_PEERS_CHANGED_ACTION:
                    outer.wifiP2pManager.requestPeers(outer.wifiP2pChannel, peers -> {
                        Collection<WifiP2pDevice> deviceList = peers.getDeviceList();
                        XLog.d("Peers changed: " + deviceList.size() + " devices");
                        if (outer.callback != null) {
                            outer.callback.onPeersChanged(deviceList);
                        }
                    });
                    break;
                    
                case WifiP2pManager.WIFI_P2P_CONNECTION_STATE_CHANGE_ACTION:
                    outer.wifiP2pManager.requestConnectionInfo(outer.wifiP2pChannel, info -> {
                        if (info.getGroupFormed()) {
                            XLog.d("Connected to group");
                            outer.connecting = false;
                            outer.connected = true;
                            outer.handler.removeCallbacks(outer.connectTimeOut);
                            if (outer.callback != null) {
                                outer.callback.onConnected(info);
                            }
                        } else {
                            XLog.d("Disconnected from group");
                            outer.connecting = false;
                            outer.connected = false;
                            if (outer.callback != null) {
                                outer.callback.onDisconnected();
                            }
                        }
                    });
                    break;
                    
                case WifiP2pManager.WIFI_P2P_THIS_DEVICE_CHANGED_ACTION:
                    WifiP2pDevice device = intent.getParcelableExtra(WifiP2pManager.EXTRA_WIFI_P2P_DEVICE);
                    if (device != null && outer.callback != null) {
                        outer.callback.onThisDeviceChanged(device);
                    }
                    break;
            }
        }
    }
    
    public static final class Companion {
        private Companion() {}
        
        public WifiP2pManagerSingleton getInstance() {
            WifiP2pManagerSingleton wifiP2pManagerSingleton = WifiP2pManagerSingleton.instance;
            if (wifiP2pManagerSingleton == null) {
                synchronized (this) {
                    wifiP2pManagerSingleton = WifiP2pManagerSingleton.instance;
                    if (wifiP2pManagerSingleton == null) {
                        Application application = GlassesWearJavaApplication.getInstance().getApplication();
                        wifiP2pManagerSingleton = new WifiP2pManagerSingleton(application);
                        WifiP2pManagerSingleton.instance = wifiP2pManagerSingleton;
                    }
                }
            }
            return wifiP2pManagerSingleton;
        }
    }
}

File: code_examples/WifiP2pManagerSingleton.kt
/**
 * WiFiP2P Manager Singleton (Kotlin Version)
 * Integrated with GlassesSDKSample's BLE approach
 * Based on: heycyan_rev/_log_review/app/src/main/java/sources/com/glasssutdio/wear/wifi/p2p/WifiP2pManagerSingleton.java
 */

package com.sdk.glassessdksample.wifip2p

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.net.wifi.p2p.WifiP2pConfig
import android.net.wifi.p2p.WifiP2pDevice
import android.net.wifi.p2p.WifiP2pInfo
import android.net.wifi.p2p.WifiP2pManager
import android.os.Handler
import android.os.Looper
import android.util.Log
import com.elvishew.xlog.XLog
import java.net.InetSocketAddress
import java.net.ServerSocket
import java.net.Socket
import java.util.concurrent.Executors

class WifiP2pManagerSingleton private constructor(private val context: Context) {
    
    companion object {
        private const val TAG = "WifiP2pManagerSingleton"
        private const val DEFAULT_PORT = 8888
        
        @Volatile
        private var instance: WifiP2pManagerSingleton? = null
        
        fun getInstance(context: Context): WifiP2pManagerSingleton {
            return instance ?: synchronized(this) {
                instance ?: WifiP2pManagerSingleton(context).also { instance = it }
            }
        }
    }
    
    // WiFiP2P components
    private val wifiP2pManager: WifiP2pManager by lazy {
        context.getSystemService(Context.WIFI_P2P_SERVICE) as WifiP2pManager
    }
    
    private var wifiP2pChannel: WifiP2pManager.Channel? = null
    private var callback: WifiP2pCallback? = null
    
    // Connection state
    private var isConnected = false
    private var isConnecting = false
    private var connectRetry = 0
    private var discoveryRetry = 0
    private var wifiP2pDevice: WifiP2pDevice? = null
    
    // Network components
    private var serverSocket: ServerSocket? = null
    private var clientSocket: Socket? = null
    private val networkExecutor = Executors.newSingleThreadExecutor()
    
    // Timeout handlers
    private val handler = Handler(Looper.getMainLooper())
    private val discoveryTimeOut = DiscoveryTimeOut(this)
    private val connectTimeOut = ConnectTimeOut(this)
    
    // Broadcast receiver
    private val receiver = WifiP2pBroadcastReceiver(this)
    private val intentFilter = IntentFilter().apply {
        addAction(WifiP2pManager.WIFI_P2P_STATE_CHANGED_ACTION)
        addAction(WifiP2pManager.WIFI_P2P_PEERS_CHANGED_ACTION)
        addAction(WifiP2pManager.WIFI_P2P_CONNECTION_STATE_CHANGE_ACTION)
        addAction(WifiP2pManager.WIFI_P2P_THIS_DEVICE_CHANGED_ACTION)
    }
    
    // Discovered peers
    val discoveredPeers = mutableListOf<WifiP2pDevice>()
    
    init {
        initializeWiFiP2P()
    }
    
    /**
     * Initialize WiFiP2P manager
     */
    private fun initializeWiFiP2P() {
        wifiP2pChannel = wifiP2pManager.initialize(context, context.mainLooper) {
            XLog.e("wifiP2pChannel disconnect")
            callback?.onDisconnected()
        }
    }
    
    /**
     * Set callback for WiFiP2P events
     */
    fun setCallback(callback: WifiP2pCallback) {
        this.callback = callback
    }
    
    /**
     * Register broadcast receiver
     */
    fun registerReceiver() {
        try {
            context.registerReceiver(receiver, intentFilter)
        } catch (e: Exception) {
            Log.e(TAG, "Error registering receiver", e)
        }
    }
    
    /**
     * Unregister broadcast receiver
     */
    fun unregisterReceiver() {
        try {
            context.unregisterReceiver(receiver)
        } catch (e: Exception) {
            Log.e(TAG, "Error unregistering receiver", e)
        }
    }
    
    /**
     * Start peer discovery
     */
    fun startPeerDiscovery() {
        if (!isConnected && !isConnecting) {
            XLog.d("Start peer discovery")
            callback?.onPeerDiscoveryStarted()
            
            wifiP2pManager.discoverPeers(wifiP2pChannel, object : WifiP2pManager.ActionListener {
                override fun onSuccess() {
                    XLog.d("Peer discovery started successfully")
                    startDiscoveryTimeout()
                }
                
                override fun onFailure(reason: Int) {
                    XLog.e("Peer discovery failed: $reason")
                    callback?.onPeerDiscoveryFailed(reason)
                    handleDiscoveryFailure()
                }
            })
        }
    }
    
    /**
     * Connect to a specific device
     */
    fun connectToDevice(device: WifiP2pDevice) {
        if (!isConnected && !isConnecting) {
            XLog.d("Connecting to device: ${device.deviceName}")
            callback?.connecting()
            
            wifiP2pDevice = device
            isConnecting = true
            
            val config = WifiP2pConfig().apply {
                deviceAddress = device.deviceAddress
            }
            
            wifiP2pManager.connect(wifiP2pChannel, config, object : WifiP2pManager.ActionListener {
                override fun onSuccess() {
                    XLog.d("Connect request sent successfully")
                    callback?.onConnectRequestSent()
                    startConnectTimeout()
                }
                
                override fun onFailure(reason: Int) {
                    XLog.e("Connect request failed: $reason")
                    callback?.onConnectRequestFailed(reason)
                    handleConnectFailure()
                }
            })
        }
    }
    
    /**
     * Cancel P2P connection
     */
    fun cancelP2pConnection() {
        if (isConnecting) {
            XLog.d("Canceling P2P connection")
            wifiP2pManager.cancelConnect(wifiP2pChannel, object : WifiP2pManager.ActionListener {
                override fun onSuccess() {
                    XLog.d("Cancel connect request sent")
                    callback?.cancelConnect()
                    resetConnectionState()
                }
                
                override fun onFailure(reason: Int) {
                    XLog.e("Cancel connect failed: $reason")
                    callback?.cancelConnectFail(reason)
                }
            })
        }
    }
    
    /**
     * Handle connection success
     */
    fun onConnectionSuccess(info: WifiP2pInfo?) {
        isConnecting = false
        isConnected = true
        connectRetry = 0
        XLog.d("WiFiP2P connected successfully")
        callback?.onConnected(info)
        
        // Start server socket for file transfer
        startServerSocket()
    }
    
    /**
     * Handle connection failure
     */
    private fun handleConnectFailure() {
        isConnecting = false
        connectRetry++
        
        if (connectRetry < MAX_CONNECT_RETRY) {
            XLog.d("Retrying connection: $connectRetry/$MAX_CONNECT_RETRY")
            handler.postDelayed({
                wifiP2pDevice?.let { connectToDevice(it) }
            }, CONNECT_RETRY_DELAY)
        } else {
            XLog.e("Max connect retries reached")
            callback?.retryAlsoFailed()
            resetConnectionState()
        }
    }
    
    /**
     * Handle discovery failure
     */
    private fun handleDiscoveryFailure() {
        discoveryRetry++
        
        if (discoveryRetry < MAX_DISCOVERY_RETRY) {
            XLog.d("Retrying discovery: $discoveryRetry/$MAX_DISCOVERY_RETRY")
            handler.postDelayed({
                startPeerDiscovery()
            }, DISCOVERY_RETRY_DELAY)
        } else {
            XLog.e("Max discovery retries reached")
            resetDiscoveryState()
        }
    }
    
    /**
     * Reset connection state
     */
    private fun resetConnectionState() {
        isConnecting = false
        isConnected = false
        wifiP2pDevice = null
        stopServerSocket()
    }
    
    /**
     * Reset discovery state
     */
    private fun resetDiscoveryState() {
        discoveryRetry = 0
        discoveredPeers.clear()
    }
    
    /**
     * Start discovery timeout
     */
    private fun startDiscoveryTimeout() {
        handler.postDelayed(discoveryTimeOut, DISCOVERY_TIMEOUT)
    }
    
    /**
     * Start connect timeout
     */
    private fun startConnectTimeout() {
        handler.postDelayed(connectTimeOut, CONNECT_TIMEOUT)
    }
    
    /**
     * Start server socket for file transfer
     */
    private fun startServerSocket() {
        networkExecutor.execute {
            try {
                serverSocket = ServerSocket(DEFAULT_PORT)
                XLog.d("Server socket started on port $DEFAULT_PORT")
                
                while (isConnected) {
                    try {
                        val client = serverSocket?.accept()
                        client?.let {
                            XLog.d("Client connected: ${it.inetAddress}")
                            handleClientConnection(it)
                        }
                    } catch (e: Exception) {
                        if (isConnected) {
                            XLog.e("Error accepting client", e)
                        }
                    }
                }
            } catch (e: Exception) {
                XLog.e("Error starting server socket", e)
            }
        }
    }
    
    /**
     * Handle client connection
     */
    private fun handleClientConnection(client: Socket) {
        // Handle file transfer communication
        try {
            // Read file transfer requests and handle accordingly
            val input = client.getInputStream()
            val buffer = ByteArray(1024)
            var bytesRead: Int
            
            while (client.isConnected && input.read(buffer).also { bytesRead = it } != -1) {
                // Process incoming data
                val data = buffer.copyOf(bytesRead)
                processReceivedData(data)
            }
        } catch (e: Exception) {
            XLog.e("Error handling client connection", e)
        } finally {
            try {
                client.close()
            } catch (e: Exception) {
                XLog.e("Error closing client socket", e)
            }
        }
    }
    
    /**
     * Process received data
     */
    private fun processReceivedData(data: ByteArray) {
        // Process file transfer commands from glasses
        XLog.d("Received data: ${data.size} bytes")
        // TODO: Implement file transfer protocol
    }
    
    /**
     * Stop server socket
     */
    private fun stopServerSocket() {
        try {
            serverSocket?.close()
            serverSocket = null
        } catch (e: Exception) {
            XLog.e("Error closing server socket", e)
        }
    }
    
    /**
     * Check if connected
     */
    fun isConnected(): Boolean = isConnected
    
    /**
     * Get discovered peers
     */
    fun getDiscoveredPeers(): List<WifiP2pDevice> = discoveredPeers.toList()
    
    /**
     * WiFiP2P Callback Interface
     */
    interface WifiP2pCallback {
        fun onPeerDiscoveryStarted()
        fun onPeerDiscoveryFailed(reason: Int)
        fun onPeersChanged(peers: Collection<WifiP2pDevice>?)
        fun onThisDeviceChanged(device: WifiP2pDevice?)
        fun onWifiP2pEnabled()
        fun onWifiP2pDisabled()
        fun onConnectRequestSent()
        fun onConnectRequestFailed(reason: Int)
        fun onConnected(info: WifiP2pInfo?)
        fun onDisconnected()
        fun connecting()
        fun cancelConnect()
        fun cancelConnectFail(reason: Int)
        fun retryAlsoFailed()
    }
    
    /**
     * Discovery timeout runnable
     */
    private class DiscoveryTimeOut(private val manager: WifiP2pManagerSingleton) : Runnable {
        override fun run() {
            XLog.d("Discovery timeout")
            manager.wifiP2pManager.stopPeerDiscovery(manager.wifiP2pChannel, null)
        }
    }
    
    /**
     * Connect timeout runnable
     */
    private class ConnectTimeOut(private val manager: WifiP2pManagerSingleton) : Runnable {
        override fun run() {
            XLog.d("Connect timeout")
            manager.handleConnectFailure()
        }
    }
    
    /**
     * Broadcast receiver for WiFiP2P events
     */
    private class WifiP2pBroadcastReceiver(private val manager: WifiP2pManagerSingleton) : BroadcastReceiver() {
        override fun onReceive(context: Context?, intent: Intent?) {
            intent?.let { handleIntent(it) }
        }
        
        private fun handleIntent(intent: Intent) {
            when (intent.action) {
                WifiP2pManager.WIFI_P2P_STATE_CHANGED_ACTION -> {
                    val state = intent.getIntExtra(WifiP2pManager.EXTRA_WIFI_STATE, -1)
                    if (state == WifiP2pManager.WIFI_P2P_STATE_ENABLED) {
                        XLog.d("WiFiP2P enabled")
                        manager.callback?.onWifiP2pEnabled()
                    } else {
                        XLog.d("WiFiP2P disabled")
                        manager.callback?.onWifiP2pDisabled()
                    }
                }
                
                WifiP2pManager.WIFI_P2P_PEERS_CHANGED_ACTION -> {
                    manager.wifiP2pManager.requestPeers(manager.wifiP2pChannel) { peers ->
                        manager.discoveredPeers.clear()
                        peers.deviceList.values.forEach { device ->
                            manager.discoveredPeers.add(device)
                        }
                        XLog.d("Peers changed: ${manager.discoveredPeers.size} devices")
                        manager.callback?.onPeersChanged(manager.discoveredPeers)
                    }
                }
                
                WifiP2pManager.WIFI_P2P_CONNECTION_STATE_CHANGE_ACTION -> {
                    manager.wifiP2pManager.requestConnectionInfo(manager.wifiP2pChannel) { info ->
                        if (info.groupFormed) {
                            XLog.d("Group formed, connected: ${info.isGroupOwner}")
                            manager.onConnectionSuccess(info)
                        } else {
                            XLog.d("Group not formed, disconnected")
                            manager.isConnected = false
                            manager.isConnecting = false
                            manager.callback?.onDisconnected()
                        }
                    }
                }
                
                WifiP2pManager.WIFI_P2P_THIS_DEVICE_CHANGED_ACTION -> {
                    val device = intent.getParcelableExtra<WifiP2pDevice>(WifiP2pManager.EXTRA_WIFI_P2P_DEVICE)
                    XLog.d("This device changed: ${device?.deviceName}")
                    manager.callback?.onThisDeviceChanged(device)
                }
            }
        }
    }
    
    companion object {
        private const val MAX_CONNECT_RETRY = 3
        private const val MAX_DISCOVERY_RETRY = 3
        private const val CONNECT_RETRY_DELAY = 2000L
        private const val DISCOVERY_RETRY_DELAY = 1000L
        private const val DISCOVERY_TIMEOUT = 30000L
        private const val CONNECT_TIMEOUT = 10000L
    }
}
